SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialRPCService.cpp

```diff
@@ -0,0 +1,474 @@
+		break;
+	}
+	case SpatialConstants::MULTICAST_RPCS_COMPONENT_ID:
+	{
+		const MulticastRPCs* Component = View->GetComponentData<MulticastRPCs>(EntityId);
+
+		if (Component->MulticastRPCBuffer.LastSentRPCId == 0 && Component->InitiallyPresentMulticastRPCsCount > 0)
+		{
+			// Update last sent ID to the number of initially present RPCs so the clients who check out this entity
+			// as it's created can process the initial multicast RPCs.
+			LastSentRPCIds.Add(EntityRPCType(EntityId, ERPCType::NetMulticast), Component->InitiallyPresentMulticastRPCsCount);
+
+			RPCRingBufferDescriptor Descriptor = RPCRingBufferUtils::GetRingBufferDescriptor(ERPCType::NetMulticast);
+			Schema_Object* SchemaObject = Schema_GetComponentUpdateFields(GetOrCreateComponentUpdate(EntityComponentId(EntityId, ComponentId)));
+			Schema_AddUint64(SchemaObject, Descriptor.LastSentRPCFieldId, Component->InitiallyPresentMulticastRPCsCount);
+		}
+		else
+		{
+			LastSentRPCIds.Add(EntityRPCType(EntityId, ERPCType::NetMulticast), Component->MulticastRPCBuffer.LastSentRPCId);
+		}
+
+		break;
+	}
+	default:
+		checkNoEntry();
+		break;
+	}
+}
+
+void SpatialRPCService::OnEndpointAuthorityLost(Worker_EntityId EntityId, Worker_ComponentId ComponentId)
+{
+	switch (ComponentId)
+	{
+	case SpatialConstants::CLIENT_ENDPOINT_COMPONENT_ID:
+	{
+		LastAckedRPCIds.Remove(EntityRPCType(EntityId, ERPCType::ClientReliable));
+		LastAckedRPCIds.Remove(EntityRPCType(EntityId, ERPCType::ClientUnreliable));
+		LastSentRPCIds.Remove(EntityRPCType(EntityId, ERPCType::ServerReliable));
+		LastSentRPCIds.Remove(EntityRPCType(EntityId, ERPCType::ServerUnreliable));
+		ClearOverflowedRPCs(EntityId);
+		break;
+	}
+	case SpatialConstants::SERVER_ENDPOINT_COMPONENT_ID:
+	{
+		LastAckedRPCIds.Remove(EntityRPCType(EntityId, ERPCType::ServerReliable));
+		LastAckedRPCIds.Remove(EntityRPCType(EntityId, ERPCType::ServerUnreliable));
+		LastSentRPCIds.Remove(EntityRPCType(EntityId, ERPCType::ClientReliable));
+		LastSentRPCIds.Remove(EntityRPCType(EntityId, ERPCType::ClientUnreliable));
+		ClearOverflowedRPCs(EntityId);
+		break;
+	}
+	case SpatialConstants::MULTICAST_RPCS_COMPONENT_ID:
+	{
+		LastSentRPCIds.Remove(EntityRPCType(EntityId, ERPCType::NetMulticast));
+		break;
+	}
+	default:
+		checkNoEntry();
+		break;
+	}
+}
+
+void SpatialRPCService::ExtractRPCsForType(Worker_EntityId EntityId, ERPCType Type)
+{
+	uint64 LastSeenRPCId;
+	EntityRPCType EntityTypePair = EntityRPCType(EntityId, Type);
+
+	if (Type == ERPCType::NetMulticast)
+	{
+		LastSeenRPCId = LastSeenMulticastRPCIds[EntityId];
+	}
+	else
+	{
+		LastSeenRPCId = LastAckedRPCIds[EntityTypePair];
+	}
+
+	const RPCRingBuffer& Buffer = GetBufferFromView(EntityId, Type);
+
+	uint64 LastProcessedRPCId = LastSeenRPCId;
+	if (Buffer.LastSentRPCId >= LastSeenRPCId)
+	{
+		for (uint64 RPCId = LastSeenRPCId + 1; RPCId <= Buffer.LastSentRPCId; RPCId++)
```

Comment:
Check that `LastSeenRPCId + 1` isn't smaller than the last overwritten one (Buffer.LastSentRPCId - buffer size)